specifications/client/verifying_client.md

```diff
@@ -0,0 +1,125 @@
+
+Links:
+
+* https://github.com/novifinancial/libra-specs/blob/master/client/client.md#updatetolatestledger-requests
+* https://github.com/novifinancial/libragogo/blob/master/client_specification.md#submittransaction-request
+* https://github.com/diem/diem/blob/master/json-rpc/json-rpc-spec.md
+
+## Data serialisation
+
+Serialisation is done using BCS (https://docs.rs/bcs/0.1.2/bcs/) it should be implemented in order to communicate with the verifying API endpoint, but also in order to verify the proofs, since these assume BCS encoded data sometimes.
+
+## Verifying signatures
+
+We refer to the [Diem Crypto spec](https://github.com/diem/diem/blob/master/specifications/crypto/README.md#signing-and-verifying) for more details regarding how the verification is to be handled. But this requires implementing BCS encoding.
+
+## JSON-RPC Endpoint
+
+The clients must query a full node, this is done using the Full Node’s JSON-RPC interface.
+Currently verification is only supported by the following (experimental) APIs:
+
+* get_state_proof
+* get_account_state_with_proof
+* get_transactions_with_proofs
+* get_events_with_proofs
+
+Notice that these endpoints should be enough to run a light client that would have fully verified proofs, without any need to trust any given full node.
+
+You can find the JSON-RPC specification here:  https://github.com/diem/diem/blob/master/json-rpc/json-rpc-spec.md
+
+## Alternative to verifying proofs
+
+Instead of verifying proofs, a client might want to :
+
+* query randomly multiple nodes and only accept an answer if it is the one given by the majority of the nodes
+* run its own full node in order to fully verify everything (this is the recommended way for VASPs, DDs and LPs)
+
+## Trust assumptions / issues
+
+If a client connects to a single full nodes and pulls all of the data from the full nodes what guarantees the client that the full nodes hasn’t been running a virtual network from Genesis onwards? → The set of initial validators is fixed and the epoch change (and thus change of validator sets) have to be signed by the previous set.
+
+## Bootstrapping
+
+There are two ways of bootstrapping a client software:
+
+* through an hardcoded Genesis and initial validator set which allow a client to verify all epoch changes since the beginning of the chain and to reach the current state of the blockchain with the assurance it has the right data. Notice that this method is subject to the so-called “long range attack” in the case where the historic validators keys are compromised, one could live in a fake fork of the network even if everything verified from the Genesis.
+* through a waypoint that represent a coarse snapshot the blockchain state at some point and which allow to catchup more easily since it doesn’t require verifying all of the epoch changes from the genesis. A given waypoint is only valid if you trust the set of validators it “contains” and if you have received the said waypoint through a trusted source.
```

Comment:
"that represents"

Reply:
"snapshot of the blockchain"

Reply:
"to catch up"

Reply:
> A given waypoint is only valid if you trust the set of validators it “contains” and if you have received the said waypoint through a trusted source.

I'm not sure what you mean here. To me the trust in a waypoint is at the source: you need to trust that the way you obtained the waypoint from a source is secure, and that the source has given you a waypoint authenticating the main branch.